### ISSUE TITLE
suppressing "I" log messages in tensorflow

### ISSUE THREAD

**Issue by davefairtex** [tensorflow/tensorflow#566]
I am using TF 0.6.0 with python 2.7.10 on both ubuntu linux and osx 10.9.5.  TF base functionality is working fine.  The issue is logging - I want to suppress the following messages that are emitted whenever I construct the first tensorflow Session:

I tensorflow/core/common_runtime/local_device.cc:40] Local device intra op parallelism threads: 2
I tensorflow/core/common_runtime/direct_session.cc:58] Direct session inter op parallelism threads: 2

They do not appear to be transmitted through the normal python logging mechanism.  Any ideas?  Did I just miss something?

I don't want to turn off all log messages, I just want TF to stop generating these two specific messages - and perhaps if you can tell me how to turn off all INFO messages from this particular (seemingly non-pythonic) source, that would be fine too.

**Comment by rekhajoshm** [comment]
Hi @davefairtex ,
I am on latest tensor, python 2.7, and osx.On a quick look, these are logged using Google logging(logging.h) in local_device.cc/direct_session.cc from common_runtime module, and these logs can be required by some users.
While invoking your application you could use flags like below to suppress as needed?
```
./your_application —minloglevel=1 (such that INFO msgs are not logged; 0/1/2/3 acceptable values defaults to 0)
./your_application —logtostderr=0 (such that log msgs are suppressed in stderr; 0/1 acceptable values)
```
Logging levels for INFO, WARNING, ERROR, and FATAL are 0, 1, 2, and 3 respectively.
Thanks!

**Comment by davefairtex** [comment]
Rekha-

Thanks much for your comment; now I at least have a clue where I should be looking.

I don't have gflags installed.  I tried using an environment variable (which has been alleged to work in comments elsewhere), but it sadly did not work for me:

ub1%  GLOG_minloglevel=3 python
Python 2.7.10 (v2.7.10:15c95b7d81dc, May 23 2015, 09:33:12) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow
>>> tensorflow.Session()
I tensorflow/core/common_runtime/local_device.cc:40] Local device intra op parallelism threads: 4
I tensorflow/core/common_runtime/direct_session.cc:58] Direct session inter op parallelism threads: 4
<tensorflow.python.client.session.Session object at 0x10693fb50>
>>> import os
>>> os.environ['GLOG_minloglevel']
'3'
>>> 

Any thoughts?  I also tried setting os.environ['GLOG_minloglevel'] = '3' within python directly prior to importing tensorflow, also no luck there.

**Comment by rekhajoshm** [comment]
Hi @davefairtex ,
From the logging referred to in tensorflow*, I observe the parameter re2::FLAGS_minloglevel that needs to be overridden. Set to 1 or greater to suppress info logs. I am not 100% sure it works as new checkins on re2 package, but Ideally it needs to be possible for user to override this parameter via command line. As a workaround, if you are stuck, you can pipeline current output to sed to remove unneeded log lines and get exact output you need.thanks.
* https://github.com/google/re2/blob/master/util/logging.h
* https://github.com/google/re2/blob/master/util/flags.h

**Comment by davefairtex** [comment]
rekha- thanks for the update.  With your help I found the re2::FLAGS_minloglevel variable in the shared object, but I have no way of setting it directly from python.  (If I were using C++ I could totally do it - but I'm not).   I tried setting it using the python gflags interface:

import tensorflow
import gflags
gflags.ADOPT_module_key_flags(tensorflow)
gflags.FLAGS(['-minloglevel', '3'])
tensorflow.Session()

but that still gives me the log messages.  Is this the correct way to set the tensorflow gflags?  (Is this actually a tensorflow module gflag?  It does show up in "nm _pywrap_tensorflow.so")

It would be awesome if the code had only one logging mechanism, controllable via python.

I know I can use sed; it just seemed so disagreeable to do that.  I was hoping there was a three-line "real" way to fix the problem.

**Comment by rekhajoshm** [comment]
Thanks @davefairtex  I agree its painful for its a simple act of suppressing and silencing logs.Was providing quick suggestions hoping simplest logical way works with flags :-) However looking more though there are various flavors of logging, this one just hits the tensorflow/core/platform/default/logging.h.For this case used verbose logging.thanks.

**Comment by rekhajoshm** [comment]
https://github.com/tensorflow/tensorflow/pull/619

**Comment by davefairtex** [comment]
"And there was great rejoicing" ...  thanks for the fix, look forward to the new version.  Its really nice to see that you guys are responsive about even minor things like this.

**Comment by fiannaca** [comment]
@davefairtex What is the actual fix? Does your glfags python code from above now work for suppressing the "I" messages?

> import tensorflow
import gflags
gflags.ADOPT_module_key_flags(tensorflow)
gflags.FLAGS(['-minloglevel', '3'])
tensorflow.Session()

**Comment by rekhajoshm** [comment]
hi @fiannaca please refer https://github.com/tensorflow/tensorflow/pull/619 for this log issue.thanks!

**Comment by fiannaca** [comment]
@rekhajoshm I should be more clear. I saw that pull request already. I'm just not sure what that means for how I can actually go about suppressing the "I" messages from TF. What are the actual steps now for suppressing log messages?

**Comment by rekhajoshm** [comment]
Hi @fiannaca As stated in the #619 , logging has a different flow here and cannot be overridden by gflags/re2 atm. You can apply the patch for the #619 for your setup to avoid the logs. Another quick pull here https://github.com/tensorflow/tensorflow/pull/1229 to have conditional logging., in some time.thanks.

**Comment by orome** [comment]
I'm also confused about the status of this. Is this possible to achieve now with a released version, or is it yet to appear?

**Comment by yaroslavvb** [comment]
There's a related issue here: https://github.com/tensorflow/tensorflow/issues/1258
That issue is still open, so I suspects there's no way to change logging settings other than modifying the code and recompiling. For instance, here are instructions on how to change verbose logging level here -- http://stackoverflow.com/a/36505898/419116

**Comment by girving** [comment]
@vrv, @rekhajoshm: What's the status of this after #619?

**Comment by vrv** [comment]
#619 just removed some unnecessary logging, our logging library still doesn't have the ability to control verbosity.  Still a contributions welcome at this point.

**Comment by davefairtex** [comment]
It has been 11 months since my initial request: a way to stifle the flood of "ALL IS STILL WELL AND OPERATING NORMALLY" info messages that appear each every time tf starts up.

Since a more general mechanism can't seem to make it onto anyone's list, can someone please just raise the default debugging level to WARN?  I don't care about INFO messages.  I suspect there is a vast silent majority that agrees with me.  Once our dev environments are configured properly, 99.9% of the problems we users face are in our own code, and the flood of (useless, to us) INFO messages proclaiming (yet again) that the normal initialization sequence worked properly just gets in the way of us trying to sort out the actual errors in our own stuff.

Thanks for listening to my rant.

Here's an example of all the stuff I have to wade through every time I run tf (0.10.0):

I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcurand.so locally
I tensorflow/core/common_runtime/gpu/gpu_init.cc:102] Found device 0 with properties: 
name: GeForce GTX 970
major: 5 minor: 2 memoryClockRate (GHz) 1.253
pciBusID 0000:01:00.0
Total memory: 3.94GiB
Free memory: 3.88GiB
I tensorflow/core/common_runtime/gpu/gpu_init.cc:126] DMA: 0 
I tensorflow/core/common_runtime/gpu/gpu_init.cc:136] 0:   Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:838] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 970, pci bus id: 0000:01:00.0)

**Comment by Fenugreek** [comment]
I am in the same boat as @davefairtex . So far I've been using a workaround: a bash script wrapping the call to my python script, filtering out the unwanted messages. Let me share it here, in case it helps someone:

```bash
#!/bin/bash
python $* 3>&1 1>&2 2>&3 3>&- | grep -v ^I\ | grep -v ^pciBusID | grep -v ^major: | grep -v ^name: |grep -v ^Total\ memory:|grep -v ^Free\ memory:
```
I called the script `tf.sh` and simpy preface python script calls with it. e.g. `tf.sh my_script.py`.

A bit non-trivial as the naive approach using `|&` and then `grep -v` ends up buffering `STDOUT`, at least on my system.

**Comment by yaroslavvb** [comment]
The more general tracking issue is https://github.com/tensorflow/tensorflow/issues/1258
I like @rekhajoshm proposal to specify logging options as part of tf.ConfigProto given to Session

**Comment by vrv** [comment]
Closing as duplicate of https://github.com/tensorflow/tensorflow/issues/1258, see comments there.

**Comment by nikhilweee** [comment]
Adding this just because I get redirected here way too often and still do not find the flag.
Terminal: `export TF_CPP_MIN_LOG_LEVEL = 2`
Python: `tf.logging.set_verbosity(tf.logging.ERROR)`
